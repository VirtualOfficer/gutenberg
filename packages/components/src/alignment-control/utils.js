/**
 * WordPress dependencies
 */
import { useState, useEffect, useRef } from '@wordpress/element';
/**
 * External dependencies
 */
import { isUndefined } from 'lodash';

export const DIRECTION = {
	UP: 'up',
	DOWN: 'down',
	LEFT: 'left',
	RIGHT: 'right',
};

export const BASE_ALIGNMENTS = [
	[ 'top', 'left' ],
	[ 'top', 'center' ],
	[ 'top', 'right' ],
	[ 'center', 'left' ],
	[ 'center' ],
	[ 'center', 'right' ],
	[ 'bottom', 'left' ],
	[ 'bottom', 'center' ],
	[ 'bottom', 'right' ],
];

export const ALIGNMENTS = BASE_ALIGNMENTS.map( normalizeAlignments );
export const ALIGNMENT_VALUES = ALIGNMENTS.map( ( a ) => a.join( ' ' ) );
export const ALIGNMENT_MATRIX = [
	[ 0, 1, 2 ],
	[ 3, 4, 5 ],
	[ 6, 7, 8 ],
];

export function useControlledState( initialState ) {
	const [ state, setState ] = useState( initialState );
	const stateRef = useRef( initialState );

	useEffect( () => {
		if ( initialState !== stateRef.current ) {
			setState( initialState );
			stateRef.current = initialState;
		}
	}, [ initialState ] );

	return [ state, setState ];
}

function normalizeAlignments( alignments = [] ) {
	const a =
		typeof alignments === 'string'
			? alignments.split( /\ |-/g )
			: alignments;
	return a.sort();
}

export function alignmentMatches( align = 'center' ) {
	const a = normalizeAlignments( align );
	return ( nextAlignment ) => {
		return JSON.stringify( nextAlignment ) === JSON.stringify( a );
	};
}

export function getAlignFromIndex( index = 0 ) {
	const align = ALIGNMENTS[ index ];
	return align.join( ' ' );
}

export function getIndexFromAlign( align = 'center' ) {
	const item = ALIGNMENTS.find( alignmentMatches( align ) );
	const index = ALIGNMENTS.indexOf( item );
	const fallbackIndex = 4; //center

	return index > -1 ? index : fallbackIndex;
}

export function getCoordsFromIndex( index = 0 ) {
	const coords = [ 0, 0 ];

	ALIGNMENT_MATRIX.forEach( ( x1, x ) => {
		x1.forEach( ( v, y ) => {
			if ( v === index ) {
				coords[ 0 ] = x;
				coords[ 1 ] = y;
			}
		} );
	} );

	return coords;
}

export function getValueFromCoords( coords, fallback = 0 ) {
	const [ x, y ] = coords;
	const exists =
		! isUndefined( ALIGNMENT_MATRIX[ x ] ) &&
		! isUndefined( ALIGNMENT_MATRIX[ x ][ y ] );
	return exists ? ALIGNMENT_MATRIX[ x ][ y ] : fallback;
}

export function getNextIndexFromDirection( currentIndex, direction ) {
	if ( ! direction ) return currentIndex;
	let nextIndex = currentIndex;
	const [ x, y ] = getCoordsFromIndex( currentIndex );

	switch ( direction ) {
		case DIRECTION.UP:
			nextIndex = getValueFromCoords( [ x - 1, y ], currentIndex );
			break;
		case DIRECTION.DOWN:
			nextIndex = getValueFromCoords( [ x + 1, y ], currentIndex );
			break;
		case DIRECTION.LEFT:
			nextIndex = getValueFromCoords( [ x, y - 1 ], currentIndex );
			break;
		case DIRECTION.RIGHT:
			nextIndex = getValueFromCoords( [ x, y + 1 ], currentIndex );
			break;
		default:
			break;
	}

	return nextIndex;
}
