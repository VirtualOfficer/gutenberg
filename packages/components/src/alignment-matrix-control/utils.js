export const GRID = [
	[ 'top left', 'top center', 'top right' ],
	[ 'center left', 'center center', 'center right' ],
	[ 'bottom left', 'bottom center', 'bottom right' ],
];

export const ALIGNMENTS = GRID.reduce( ( alignments, row ) => {
	row.forEach( ( cell ) => {
		alignments.push( cell );
	} );

	return alignments;
}, [] );

export function parseValue( value ) {
	const nextValue = value === 'center' ? 'center center' : value;

	return nextValue.replace( '-', ' ' );
}

export function getItemId( id, value ) {
	const valueId = parseValue( value ).replace( '-', ' ' );

	return `${ id }-${ valueId }`;
}

/**
 * Retrieves the [x, y] alignment data index from a value.
 *
 * @param {string} alignment Value to check.
 *
 * @return {number} The index of a matching alignment.
 */
export function getAlignmentIndex( alignment = 'center' ) {
	const item = parseValue( alignment ).replace( '-', ' ' );
	const index = ALIGNMENTS.indexOf( item );

	return index > -1 ? index : undefined;
}
