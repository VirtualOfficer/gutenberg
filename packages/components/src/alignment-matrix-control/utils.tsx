/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import type {
	AlignmentMatrixControlValue,
	VoidableAlignmentMatrixControlValue,
} from './types';

export const GRID: AlignmentMatrixControlValue[][] = [
	[ 'top left', 'top center', 'top right' ],
	[ 'center left', 'center center', 'center right' ],
	[ 'bottom left', 'bottom center', 'bottom right' ],
];

// Stored as map as i18n __() only accepts strings (not variables)
export const ALIGNMENT_LABEL: Record< AlignmentMatrixControlValue, string > = {
	'top left': __( 'Top Left' ),
	'top center': __( 'Top Center' ),
	'top right': __( 'Top Right' ),
	'center left': __( 'Center Left' ),
	'center center': __( 'Center' ),
	center: __( 'Center' ),
	'center right': __( 'Center Right' ),
	'bottom left': __( 'Bottom Left' ),
	'bottom center': __( 'Bottom Center' ),
	'bottom right': __( 'Bottom Right' ),
};

// Transforms GRID into a flat Array of values.
export const ALIGNMENTS = GRID.flat();

/**
 * Normalizes an incoming value to better match the alignment values
 *
 * @param value an alignment value to normalize
 *
 * @return The normalized value
 */
export function normalizeValue( value: VoidableAlignmentMatrixControlValue ) {
	return value === 'center' ? 'center center' : value;
}

/**
 * Normalizes and transforms an incoming value to better match the alignment values
 *
 * @param value An alignment value to parse.
 *
 * @return The parsed value.
 */
export function transformValue( value: VoidableAlignmentMatrixControlValue ) {
	return normalizeValue( value )?.replace(
		'-',
		' '
	) as VoidableAlignmentMatrixControlValue;
}

/**
 * Creates an item ID based on a prefix ID and an alignment value.
 *
 * @param prefixId An ID to prefix.
 * @param value    An alignment value.
 *
 * @return The item id.
 */
export function getItemId(
	prefixId: string,
	value: VoidableAlignmentMatrixControlValue
) {
	const valueId = transformValue( value )?.replace( ' ', '-' );
	return valueId && `${ prefixId }-${ valueId }`;
}

/**
 * Extracts an item value from its ID
 *
 * @param prefixId An ID prefix to remove
 * @param id       An item ID
 * @return         The item value
 */
export function getItemValue(
	prefixId: string,
	id: VoidableAlignmentMatrixControlValue
) {
	return transformValue(
		id?.replace( prefixId + '-', '' ) as VoidableAlignmentMatrixControlValue
	);
}

/**
 * Retrieves the alignment index from a value.
 *
 * @param alignment Value to check.
 *
 * @return The index of a matching alignment.
 */
export function getAlignmentIndex(
	alignment: VoidableAlignmentMatrixControlValue = 'center'
) {
	const transformedValue = transformValue(
		alignment
	) as AlignmentMatrixControlValue;

	// `transformedValue` could come back as undefined or null,
	// but `indexOf` will still return -1, so we can proceed
	// without any problems.
	const index = ALIGNMENTS.indexOf( transformedValue );

	return index > -1 ? index : undefined;
}
