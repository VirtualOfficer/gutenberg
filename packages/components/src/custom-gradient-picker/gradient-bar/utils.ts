// @ts-nocheck

/**
 * Internal dependencies
 */
import { MINIMUM_DISTANCE_BETWEEN_POINTS } from './constants';
import type { ControlPoint } from '../types';

export function clampPercent( value: number ) {
	return Math.max( 0, Math.min( 100, value ) );
}

export function isOverlapping(
	value: ControlPoint[],
	initialIndex: number,
	newPosition: number,
	minDistance = MINIMUM_DISTANCE_BETWEEN_POINTS
) {
	const initialPosition = value[ initialIndex ].position;
	const minPosition = Math.min( initialPosition, newPosition );
	const maxPosition = Math.max( initialPosition, newPosition );

	return value.some( ( { position }, index ) => {
		return (
			index !== initialIndex &&
			( Math.abs( position - newPosition ) < minDistance ||
				( minPosition < position && position < maxPosition ) )
		);
	} );
}

export function addControlPoint(
	points: ControlPoint[],
	position: number,
	color: string
) {
	const nextIndex = points.findIndex(
		( point ) => point.position > position
	);
	const newPoint = { color, position };
	const newPoints = points.slice();
	newPoints.splice( nextIndex - 1, 0, newPoint );
	return newPoints;
}

export function removeControlPoint( points: ControlPoint[], index: number ) {
	return points.filter( ( _point, pointIndex ) => {
		return pointIndex !== index;
	} );
}

export function updateControlPoint(
	points: ControlPoint[],
	index: number,
	newPoint: ControlPoint
) {
	const newValue = points.slice();
	newValue[ index ] = newPoint;
	return newValue;
}

export function updateControlPointPosition(
	points: ControlPoint[],
	index: number,
	newPosition: ControlPoint[ 'position' ]
) {
	if ( isOverlapping( points, index, newPosition ) ) {
		return points;
	}
	const newPoint = {
		...points[ index ],
		position: newPosition,
	};
	return updateControlPoint( points, index, newPoint );
}

export function updateControlPointColor(
	points: ControlPoint[],
	index: number,
	newColor: string
) {
	const newPoint = {
		...points[ index ],
		color: newColor,
	};
	return updateControlPoint( points, index, newPoint );
}

export function updateControlPointColorByPosition(
	points: ControlPoint[],
	position: ControlPoint[ 'position' ],
	newColor: string
) {
	const index = points.findIndex( ( point ) => point.position === position );
	return updateControlPointColor( points, index, newColor );
}

export function getHorizontalRelativeGradientPosition(
	mouseXCoordinate: number,
	containerElement: HTMLDivElement | null
) {
	if ( ! containerElement ) {
		return;
	}
	const { x, width } = containerElement.getBoundingClientRect();
	const absolutePositionValue = mouseXCoordinate - x;
	return Math.round(
		clampPercent( ( absolutePositionValue * 100 ) / width )
	);
}
