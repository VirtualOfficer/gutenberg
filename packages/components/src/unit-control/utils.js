export const CSS_UNITS = [
	{ value: 'px', label: 'px' },
	{ value: '%', label: '%' },
	{ value: 'em', label: 'em' },
	{ value: 'rem', label: 'rem' },
	{ value: 'vw', label: 'vw' },
	{ value: 'vh', label: 'vh' },
];
