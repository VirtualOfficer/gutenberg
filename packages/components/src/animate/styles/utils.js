/**
 * Internal dependencies
 */
import { getRTL } from '../../utils';

export const getRTLSafeTranslateX = ( { origin } ) => {
	const isRTL = getRTL();

	if ( isRTL ) {
		return origin === 'right' ? '-100%' : '+100%';
	}

	return origin === 'right' ? '+100%' : '-100%';
};

export const getRTLSafeXAxis = ( { xAxis } ) => {
	const isRTL = getRTL();

	if ( isRTL ) {
		switch ( xAxis ) {
			case 'left':
				return 'right';
			case 'right':
				return 'left';
			default:
				return xAxis;
		}
	}

	return xAxis;
};
