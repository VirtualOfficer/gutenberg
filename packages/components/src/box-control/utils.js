/**
 * External dependencies
 */
import { isEmpty, isNumber } from 'lodash';

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { parseUnit } from '../unit-control/utils';

export const LABELS = {
	all: __( 'All' ),
	top: __( 'Top' ),
	bottom: __( 'Bottom' ),
	left: __( 'Left' ),
	right: __( 'Right' ),
	mixed: __( 'Mixed' ),
};

export const CUSTOM_VALUES = {
	AUTO: 'auto',
};

export const DEFAULT_VALUES = {
	top: null,
	right: null,
	bottom: null,
	left: null,
};

export const DEFAULT_VISUALIZER_VALUES = {
	top: false,
	right: false,
	bottom: false,
	left: false,
};

export const DEFAULT_VISUALIZER_SPACING_VALUES = {
	padding: DEFAULT_VISUALIZER_VALUES,
	margin: DEFAULT_VISUALIZER_VALUES,
};

/**
 * Gets an items with the most occurance within an array
 * https://stackoverflow.com/a/20762713
 *
 * @param {Array<any>} arr Array of items to check.
 * @return {any} The item with the most occurances.
 */
function mode( arr ) {
	return arr
		.sort(
			( a, b ) =>
				arr.filter( ( v ) => v === a ).length -
				arr.filter( ( v ) => v === b ).length
		)
		.pop();
}

/**
 * Gets the 'all' input value and unit from values data.
 *
 * @param {Object} values Box values.
 * @return {string} A value + unit for the 'all' input.
 */
export function getAllValue( values = {} ) {
	const parsedValues = Object.values( values ).map( ( value ) =>
		parseUnit( value )
	);

	const allValues = parsedValues.map( ( value ) => value[ 0 ] );
	const allUnits = parsedValues.map( ( value ) => value[ 1 ] );

	const value = allValues.every( ( v ) => v === allValues[ 0 ] )
		? allValues[ 0 ]
		: '';
	const unit = mode( allUnits );

	/**
	 * The isNumber check is important. On reset actions, the incoming value
	 * may be null or an empty string.
	 *
	 * Also, the value may also be zero (0), which is considered a valid unit value.
	 *
	 * isNumber() is more specific for these cases, rather than relying on a
	 * simple truthy check.
	 */
	const allValue = isNumber( value ) ? `${ value }${ unit }` : null;
	return allValue;
}

/**
 * Checks to determine if values are mixed.
 *
 * @param {Object} values Box values.
 * @return {boolean} Whether values are mixed.
 */
export function isValuesMixed( values = {} ) {
	const vals = Object.values( values );
	return vals.some( ( val ) => val !== vals[ 0 ] );
}

/**
 * Checks to determine if values are defined.
 *
 * @param {Object} values Box values.
 *
 * @return {boolean} Whether values are mixed.
 */
export function isValuesDefined( values ) {
	return (
		values !== undefined &&
		! isEmpty( Object.values( values ).filter( Boolean ) )
	);
}

const sideStyles = {
	margin: {
		top: { transform: 'translateY(-100%)' },
		right: { transform: 'translateX(100%)' },
		bottom: { transform: 'translateY(100%)' },
		left: { transform: 'translateX(-100%)' },
	},
	padding: {
		top: null,
		right: null,
		bottom: null,
		left: null,
	},
};

/**
 * Modifies the style properties of each side.
 *
 * @param {string} type of field
 * @return {function(*, [*, *]): *} reducer function adding additional styles
 */

export function extendStyles( type ) {
	return ( acc, [ side, value ] ) => {
		const styles = sideStyles[ type ][ side ];
		return {
			...acc,
			[ side ]: {
				style: {
					...styles,
					[ side ]: value,
				},
			},
		};
	};
}
