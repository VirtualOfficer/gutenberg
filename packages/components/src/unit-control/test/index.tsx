/**
 * External dependencies
 */
import { describe, expect, it } from '@jest/globals';
import { render, fireEvent, screen } from '@testing-library/react';

/**
 * WordPress dependencies
 */
import { UP, DOWN, ENTER, ESCAPE } from '@wordpress/keycodes';
import { useState } from '@wordpress/element';

/**
 * Internal dependencies
 */
import UnitControl from '../';
import { parseQuantityAndUnitFromRawValue } from '../utils';
import type { UnitControlOnChangeCallback } from '../types';

// TODO: What determines spinbutton or textbox ?
const getInput = ( isNumeric: boolean = false ) =>
	screen.getByRole(
		isNumeric ? 'spinbutton' : 'textbox'
	) as HTMLInputElement;
const getSelect = () => screen.getByRole( 'combobox' ) as HTMLSelectElement;
const getSelectOptions = () =>
	screen.getAllByRole( 'option' ) as HTMLOptionElement[];

const ControlledSyncUnits = () => {
	const [ state, setState ] = useState( {
		valueA: '',
		valueB: '',
	} );

	// Keep the unit sync'd between the two `UnitControl` instances.
	const onUnitControlChange = (
		fieldName: 'valueA' | 'valueB',
		newValue?: string | number
	) => {
		const parsedQuantityAndUnit = parseQuantityAndUnitFromRawValue(
			newValue
		);
		const quantity = parsedQuantityAndUnit[ 0 ];

		if ( ! Number.isFinite( quantity ) ) {
			return;
		}

		const newUnit = parsedQuantityAndUnit[ 1 ];

		const nextState = {
			...state,
			[ fieldName ]: newValue,
		};

		Object.entries( state ).forEach( ( [ stateProp, stateValue ] ) => {
			const [
				stateQuantity,
				stateUnit,
			] = parseQuantityAndUnitFromRawValue( stateValue );

			if ( stateProp !== fieldName && stateUnit !== newUnit ) {
				nextState[
					stateProp as 'valueA' | 'valueB'
				] = `${ stateQuantity }${ newUnit }`;
			}
		} );

		setState( nextState );
	};

	return (
		<>
			<UnitControl
				label="Field A"
				value={ state.valueA }
				onChange={ ( v ) => onUnitControlChange( 'valueA', v ) }
			/>
			<UnitControl
				label="Field B"
				value={ state.valueB }
				onChange={ ( v ) => onUnitControlChange( 'valueB', v ) }
			/>
		</>
	);
};

describe( 'UnitControl', () => {
	describe( 'Basic rendering', () => {
		it( 'should render', () => {
			render( <UnitControl /> );
			const input = getInput( true );
			const select = getSelect();

			expect( input ).toBeInTheDocument();
			expect( select ).toBeInTheDocument();
		} );

		// TODO: update TS matchers
		it( 'should render custom className', () => {
			const { container: noClassName } = render( <UnitControl /> );

			const { container: withClassName } = render(
				<UnitControl className="hello" />
			);

			expect( noClassName.firstChild ).toMatchDiffSnapshot(
				withClassName.firstChild
			);
		} );

		// TODO: update TS matchers
		it( 'should not render select, if units are disabled', () => {
			render( <UnitControl unit="em" units={ [] } /> );
			const input = getInput( true );
			const select = screen.queryByRole( 'combobox' );

			expect( input ).toBeInTheDocument();
			expect( select ).not.toBeInTheDocument();
		} );

		// TODO: update TS matchers
		// Check why it errors
		it.skip( 'should render label if single units', () => {
			render( <UnitControl units={ [ { value: '%', label: '%' } ] } /> );

			const select = screen.queryByRole( 'combobox' );
			const label = screen.getByText( '%' );

			expect( select ).not.toBeInTheDocument();
			expect( label ).toBeInTheDocument();
		} );
	} );

	describe( 'Value', () => {
		it( 'should update value on change', () => {
			let state = '50px';
			const setState = jest.fn( ( value ) => ( state = value ) );

			render( <UnitControl value={ state } onChange={ setState } /> );

			const input = getInput( true );
			input.focus();
			fireEvent.change( input, { target: { value: 62 } } );

			expect( setState ).toHaveBeenCalledTimes( 1 );
			expect( state ).toBe( '62px' );
		} );

		it( 'should increment value on UP press', () => {
			let state: string | undefined = '50px';
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			render( <UnitControl value={ state } onChange={ setState } /> );

			const input = getInput( true );
			input.focus();
			fireEvent.keyDown( input, { keyCode: UP } );

			expect( state ).toBe( '51px' );
		} );

		it( 'should increment value on UP + SHIFT press, with step', () => {
			let state: string | undefined = '50px';
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			render( <UnitControl value={ state } onChange={ setState } /> );

			const input = getInput( true );
			input.focus();
			fireEvent.keyDown( input, { keyCode: UP, shiftKey: true } );

			expect( state ).toBe( '60px' );
		} );

		it( 'should decrement value on DOWN press', () => {
			let state: string | number | undefined = 50;
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			render( <UnitControl value={ state } onChange={ setState } /> );

			const input = getInput( true );
			input.focus();
			fireEvent.keyDown( input, { keyCode: DOWN } );

			expect( state ).toBe( '49px' );
		} );

		it( 'should decrement value on DOWN + SHIFT press, with step', () => {
			let state: string | number | undefined = 50;
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			render( <UnitControl value={ state } onChange={ setState } /> );

			const input = getInput( true );
			input.focus();
			fireEvent.keyDown( input, { keyCode: DOWN, shiftKey: true } );

			expect( state ).toBe( '40px' );
		} );

		it( 'should cancel change when ESCAPE key is pressed', () => {
			let state: string | number | undefined = 50;
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					isPressEnterToChange
				/>
			);

			const input = getInput() as HTMLInputElement;
			input.focus();

			fireEvent.change( input, { target: { value: '300px' } } );

			expect( input.value ).toBe( '300px' );
			expect( state ).toBe( 50 );

			fireEvent.keyDown( input, { keyCode: ESCAPE } );

			expect( input.value ).toBe( '50' );
			expect( state ).toBe( 50 );
		} );
	} );

	describe( 'Unit', () => {
		it( 'should update unit value on change', () => {
			let state: string | undefined = 'px';
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			render( <UnitControl unit={ state } onUnitChange={ setState } /> );

			const select = getSelect();
			select.focus();
			fireEvent.change( select, { target: { value: 'em' } } );

			expect( state ).toBe( 'em' );
		} );

		it( 'should render customized units, if defined', () => {
			const units = [
				{ value: 'pt', label: 'pt', default: 0 },
				{ value: 'vmax', label: 'vmax', default: 10 },
			];

			render( <UnitControl units={ units } /> );

			const options = getSelectOptions();

			expect( options.length ).toBe( 2 );

			const [ pt, vmax ] = options;

			expect( pt.value ).toBe( 'pt' );
			expect( vmax.value ).toBe( 'vmax' );
		} );

		it( 'should reset value on unit change, if unit has default value', () => {
			let state: string | number | undefined = 50;
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			const units = [
				{ value: 'pt', label: 'pt', default: 25 },
				{ value: 'vmax', label: 'vmax', default: 75 },
			];

			render(
				<UnitControl
					isResetValueOnUnitChange
					units={ units }
					onChange={ setState }
					value={ state }
				/>
			);

			const select = getSelect();
			select.focus();

			fireEvent.change( select, { target: { value: 'vmax' } } );

			expect( state ).toBe( '75vmax' );

			fireEvent.change( select, { target: { value: 'pt' } } );

			expect( state ).toBe( '25pt' );
		} );

		it( 'should not reset value on unit change, if disabled', () => {
			let state: string | number | undefined = 50;
			const setState: UnitControlOnChangeCallback = ( nextState ) =>
				( state = nextState );

			const units = [
				{ value: 'pt', label: 'pt', default: 25 },
				{ value: 'vmax', label: 'vmax', default: 75 },
			];

			render(
				<UnitControl
					isResetValueOnUnitChange={ false }
					value={ state }
					units={ units }
					onChange={ setState }
				/>
			);

			const select = getSelect();
			select.focus();

			fireEvent.change( select, { target: { value: 'vmax' } } );

			expect( state ).toBe( '50vmax' );

			fireEvent.change( select, { target: { value: 'pt' } } );

			expect( state ).toBe( '50pt' );
		} );

		it( 'should set correct unit if single units', () => {
			let state: string | undefined = '50%';
			const setState: UnitControlOnChangeCallback = ( value ) =>
				( state = value );

			render(
				<UnitControl
					value={ state }
					unit="%"
					units={ [ { value: '%', label: '%' } ] }
					onChange={ setState }
				/>
			);

			const input = getInput( true );
			input.focus();
			fireEvent.change( input, { target: { value: 62 } } );

			expect( state ).toBe( '62%' );
		} );

		// TODO: update TS matchers
		it( 'should update unit value when a new raw value is passed', () => {
			render( <ControlledSyncUnits /> );

			const [ inputA, inputB ] = screen.getAllByRole( 'spinbutton' );
			const [ selectA, selectB ] = screen.getAllByRole( 'combobox' );

			inputA.focus();
			fireEvent.change( inputA, { target: { value: '55' } } );

			inputB.focus();
			fireEvent.change( inputB, { target: { value: '14' } } );

			selectA.focus();
			fireEvent.change( selectA, { target: { value: 'rem' } } );

			expect( selectA ).toHaveValue( 'rem' );
			expect( selectB ).toHaveValue( 'rem' );

			selectB.focus();
			fireEvent.change( selectB, { target: { value: 'vw' } } );

			expect( selectA ).toHaveValue( 'vw' );
			expect( selectB ).toHaveValue( 'vw' );
		} );
	} );

	describe( 'Unit Parser', () => {
		let state = '10px';
		const setState = jest.fn( ( nextState ) => ( state = nextState ) );

		it( 'should parse unit from input', () => {
			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					isPressEnterToChange
				/>
			);

			const input = getInput();
			input.focus();
			fireEvent.change( input, { target: { value: '55 em' } } );
			fireEvent.keyDown( input, { keyCode: ENTER } );

			expect( state ).toBe( '55em' );
		} );

		it( 'should parse PX unit from input', () => {
			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					isPressEnterToChange
				/>
			);

			const input = getInput();
			input.focus();
			fireEvent.change( input, { target: { value: '61   PX' } } );
			fireEvent.keyDown( input, { keyCode: ENTER } );

			expect( state ).toBe( '61px' );
		} );

		it( 'should parse EM unit from input', () => {
			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					isPressEnterToChange
				/>
			);

			const input = getInput();
			input.focus();
			fireEvent.change( input, { target: { value: '55 em' } } );
			fireEvent.keyDown( input, { keyCode: ENTER } );

			expect( state ).toBe( '55em' );
		} );

		it( 'should parse % unit from input', () => {
			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					isPressEnterToChange
				/>
			);

			const input = getInput();
			input.focus();
			fireEvent.change( input, { target: { value: '-10  %' } } );
			fireEvent.keyDown( input, { keyCode: ENTER } );

			expect( state ).toBe( '-10%' );
		} );

		it( 'should parse REM unit from input', () => {
			render(
				<UnitControl
					value={ state }
					onChange={ setState }
					isPressEnterToChange
				/>
			);

			const input = getInput();
			input.focus();
			fireEvent.change( input, {
				target: { value: '123       rEm  ' },
			} );
			fireEvent.keyDown( input, { keyCode: ENTER } );

			expect( state ).toBe( '123rem' );
		} );

		it( 'should update unit after initial render and with new unit prop', () => {
			const { rerender } = render( <UnitControl value={ '10%' } /> );

			const select = getSelect() as HTMLSelectElement;

			expect( select.value ).toBe( '%' );

			rerender( <UnitControl value={ '20' } unit="em" /> );

			expect( select.value ).toBe( 'em' );
		} );

		it( 'should fallback to default unit if parsed unit is invalid', () => {
			render( <UnitControl value={ '10null' } /> );

			const select = getSelect() as HTMLSelectElement;
			expect( select.value ).toBe( 'px' );
		} );

		it( 'should display valid CSS unit when not explicitly included in units list', () => {
			render(
				<UnitControl
					value={ '10%' }
					units={ [
						{ value: 'px', label: 'px' },
						{ value: 'em', label: 'em' },
					] }
				/>
			);

			const select = getSelect();
			const options = getSelectOptions();

			expect( select.value ).toBe( '%' );
			expect( options.length ).toBe( 3 );
		} );
	} );
} );
