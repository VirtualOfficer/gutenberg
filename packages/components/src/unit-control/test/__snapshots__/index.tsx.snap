// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnitControl Basic rendering should render custom className 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -1,18 +1,18 @@
  <div>
    <div
-     class="components-flex components-input-control components-number-control components-unit-control components-unit-control-wrapper e1bagdl32 ep09it41 em5sgkm7 css-zruo0h-View-Flex-base-ItemsColumn-Root-Input-htmlArrowStyles-ValueInput e19lxcc00"
+     class="components-flex components-input-control components-number-control components-unit-control components-unit-control-wrapper hello e1bagdl32 ep09it41 em5sgkm7 css-zruo0h-View-Flex-base-ItemsColumn-Root-Input-htmlArrowStyles-ValueInput e19lxcc00"
      data-wp-c16t="true"
      data-wp-component="Flex"
    >
      <div
        class="components-input-control__container css-1o1a8cj-Container-containerDisabledStyles-containerWidthStyles em5sgkm6"
      >
        <input
          autocomplete="off"
          class="components-input-control__input css-1hsyo1w-Input-dragStyles-fontSizeStyles-sizeStyles-customPaddings em5sgkm5"
-         id="inspector-input-control-1"
+         id="inspector-input-control-2"
          inputmode="numeric"
          max="Infinity"
          min="-Infinity"
          step="1"
          type="number"
`;
